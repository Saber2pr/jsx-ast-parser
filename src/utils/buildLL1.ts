import { buildLexer } from 'typescript-parsec'

export const buildLL1: (words: string[], kind: any) => typeof buildLexer =
  (words, kind) => rules => {
    const conflicts: any[][] = []
    for (const word of words) {
      const matched = []
      for (const rule of rules) {
        if (word.match(rule[1])) {
          matched.push(kind[rule[2]])
        }
      }
      if (matched.length > 1) {
        conflicts.push([word, ...matched])
      }
    }
    if (conflicts.length) {
      console.log(
        `[LL1 Conflicts] \n${conflicts
          .map(row => `${row[0]}: ${row.slice(1).join(',')}`)
          .join('\n')}`
      )
      process.exit(1)
    }
    return buildLexer(rules)
  }
