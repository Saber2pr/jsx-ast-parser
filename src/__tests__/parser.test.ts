/*
 * @Author: saber2pr
 * @Date: 2021-09-12 12:06:08
 * @Last Modified by: saber2pr
 * @Last Modified time: 2021-10-07 11:07:05
 */
import { parser } from '../'
import { read } from './utils'

describe('Parser', () => {
  it('Case 1', async () => {
    const code = await read('code.txt')
    expect(parser.parse(code)).toMatchSnapshot()
  })

  it('CallChainExpr', () => {
    expect(
      parser.isCallChainExpr(parser.parse('console.log(error)').body[0])
    ).toBe(true)
  })

  it('DefineVariableStatement', () => {
    expect(
      parser.isDefineVariableStatement(parser.parse('const a = "a"').body[0])
    ).toBe(true)
  })

  it('VariableAssignExpr', () => {
    expect(parser.isVariableAssignExpr(parser.parse('a = "a"').body[0])).toBe(
      true
    )
  })

  it('IfStatement', () => {
    expect(
      parser.isIfStatement(
        parser.parse(`
    if(a){
      console.log(a)
    } else if(b) {
      console.log(b)
    } else {
      console.log(c)
    }
    `).body[0]
      )
    ).toBe(true)
  })

  it('ReturnStatement', () => {
    expect(parser.isReturnStatement(parser.parse('return 233').body[0])).toBe(
      true
    )
  })

  it('JsxExpr', () => {
    expect(parser.isJsxExpr(parser.parse('<div>233</div>').body[0])).toBe(true)
  })

  it('JsxSelfClosingExpr', () => {
    expect(
      parser.isJsxSelfClosingExpr(
        parser.parse('<div className="233" />').body[0]
      )
    ).toBe(true)
  })

  it('StringExpr', () => {
    expect(parser.isStringExpr(parser.parse('"saber2pr"').body[0])).toBe(true)
  })

  it('NumberExpr', () => {
    expect(parser.isNumberExpr(parser.parse('233').body[0])).toBe(true)
  })

  it('ObjectExpr', () => {
    expect(parser.isObjectExpr(parser.parse('{a:"a"}').body[0])).toBe(true)
  })

  it('ArrayExpr', () => {
    expect(parser.isArrayExpr(parser.parse('["a", "b"]').body[0])).toBe(true)
  })

  it('ArrowFunctionExpr', () => {
    expect(
      parser.isArrowFunctionExpr(
        parser.parse('(a) => { console.log(a) }').body[0]
      )
    ).toBe(true)
  })

  it('FunctionExpr', () => {
    expect(
      parser.isFunctionExpr(
        parser.parse('function test(a){ console.log(a) }').body[0]
      )
    ).toBe(true)
  })
})
