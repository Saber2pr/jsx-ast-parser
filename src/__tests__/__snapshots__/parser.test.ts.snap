// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser Case 1 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "kind": "JsxSelfClosingExpr",
          "props": Array [
            Object {
              "key": Object {
                "kind": "IdentityExpr",
                "name": "list",
              },
              "kind": "PropExpr",
              "value": Object {
                "items": Array [
                  Object {
                    "kind": "ObjectExpr",
                    "props": Object {
                      "content": Object {
                        "body": Array [
                          Object {
                            "kind": "TextExpr",
                            "value": "233",
                          },
                        ],
                        "closingTag": Object {
                          "kind": "ClosingTagExpr",
                          "tagName": Object {
                            "kind": "IdentityExpr",
                            "name": "View",
                          },
                        },
                        "kind": "JsxExpr",
                        "openingTag": Object {
                          "kind": "OpeningTagExpr",
                          "props": Array [
                            Object {
                              "key": Object {
                                "kind": "IdentityExpr",
                                "name": "color",
                              },
                              "kind": "PropExpr",
                              "value": Object {
                                "kind": "StringExpr",
                                "value": "red",
                              },
                            },
                          ],
                          "tagName": Object {
                            "kind": "IdentityExpr",
                            "name": "View",
                          },
                        },
                      },
                      "logo": Object {
                        "kind": "JsxSelfClosingExpr",
                        "props": Array [
                          Object {
                            "key": Object {
                              "kind": "IdentityExpr",
                              "name": "mode",
                            },
                            "kind": "PropExpr",
                            "value": Object {
                              "kind": "StringExpr",
                              "value": "test",
                            },
                          },
                        ],
                        "tagName": Object {
                          "kind": "IdentityExpr",
                          "name": "Image",
                        },
                      },
                    },
                  },
                  Object {
                    "kind": "ObjectExpr",
                    "props": Object {
                      "content": Object {
                        "kind": "JsxSelfClosingExpr",
                        "props": Array [],
                        "tagName": Object {
                          "kind": "IdentityExpr",
                          "name": "View",
                        },
                      },
                    },
                  },
                ],
                "kind": "ArrayExpr",
              },
            },
          ],
          "tagName": Object {
            "kind": "IdentityExpr",
            "name": "List",
          },
        },
        Object {
          "kind": "JsxSelfClosingExpr",
          "props": Array [],
          "tagName": Object {
            "kind": "IdentityExpr",
            "name": "div",
          },
        },
        Object {
          "kind": "JsxSelfClosingExpr",
          "props": Array [
            Object {
              "key": Object {
                "kind": "IdentityExpr",
                "name": "id",
              },
              "kind": "PropExpr",
              "value": Object {
                "kind": "StringExpr",
                "value": "qwq",
              },
            },
          ],
          "tagName": Object {
            "kind": "IdentityExpr",
            "name": "div",
          },
        },
        Object {
          "body": Array [
            Object {
              "kind": "TextExpr",
              "value": "aaa",
            },
          ],
          "closingTag": Object {
            "kind": "ClosingTagExpr",
            "tagName": Object {
              "kind": "IdentityExpr",
              "name": "span",
            },
          },
          "kind": "JsxExpr",
          "openingTag": Object {
            "kind": "OpeningTagExpr",
            "props": Array [],
            "tagName": Object {
              "kind": "IdentityExpr",
              "name": "span",
            },
          },
        },
        Object {
          "body": Array [
            Object {
              "kind": "TextExpr",
              "value": "1234",
            },
          ],
          "closingTag": Object {
            "kind": "ClosingTagExpr",
            "tagName": Object {
              "kind": "IdentityExpr",
              "name": "span",
            },
          },
          "kind": "JsxExpr",
          "openingTag": Object {
            "kind": "OpeningTagExpr",
            "props": Array [],
            "tagName": Object {
              "kind": "IdentityExpr",
              "name": "span",
            },
          },
        },
        Object {
          "body": Array [
            Object {
              "kind": "TextExpr",
              "value": "1234asd",
            },
          ],
          "closingTag": Object {
            "kind": "ClosingTagExpr",
            "tagName": Object {
              "kind": "IdentityExpr",
              "name": "span",
            },
          },
          "kind": "JsxExpr",
          "openingTag": Object {
            "kind": "OpeningTagExpr",
            "props": Array [],
            "tagName": Object {
              "kind": "IdentityExpr",
              "name": "span",
            },
          },
        },
        Object {
          "body": Array [
            Object {
              "kind": "TextExpr",
              "value": "12aa",
            },
            Object {
              "body": Array [
                Object {
                  "kind": "TextExpr",
                  "value": "aaa",
                },
              ],
              "closingTag": Object {
                "kind": "ClosingTagExpr",
                "tagName": Object {
                  "kind": "IdentityExpr",
                  "name": "span",
                },
              },
              "kind": "JsxExpr",
              "openingTag": Object {
                "kind": "OpeningTagExpr",
                "props": Array [],
                "tagName": Object {
                  "kind": "IdentityExpr",
                  "name": "span",
                },
              },
            },
            Object {
              "kind": "TextExpr",
              "value": "aa234234aaa",
            },
          ],
          "closingTag": Object {
            "kind": "ClosingTagExpr",
            "tagName": Object {
              "kind": "IdentityExpr",
              "name": "span",
            },
          },
          "kind": "JsxExpr",
          "openingTag": Object {
            "kind": "OpeningTagExpr",
            "props": Array [],
            "tagName": Object {
              "kind": "IdentityExpr",
              "name": "span",
            },
          },
        },
      ],
      "closingTag": Object {
        "kind": "ClosingTagExpr",
        "tagName": Object {
          "kind": "IdentityExpr",
          "name": "div",
        },
      },
      "kind": "JsxExpr",
      "openingTag": Object {
        "kind": "OpeningTagExpr",
        "props": Array [
          Object {
            "key": Object {
              "kind": "IdentityExpr",
              "name": "width",
            },
            "kind": "PropExpr",
            "value": Object {
              "kind": "NumberExpr",
              "value": 100,
            },
          },
          Object {
            "key": Object {
              "kind": "IdentityExpr",
              "name": "contentEditable",
            },
            "kind": "PropExpr",
            "value": Object {
              "kind": "BooleanExpr",
              "value": true,
            },
          },
          Object {
            "key": Object {
              "kind": "IdentityExpr",
              "name": "color",
            },
            "kind": "PropExpr",
            "value": Object {
              "kind": "StringExpr",
              "value": "red",
            },
          },
          Object {
            "key": Object {
              "kind": "IdentityExpr",
              "name": "style",
            },
            "kind": "PropExpr",
            "value": Object {
              "kind": "ObjectExpr",
              "props": Object {
                "background": Object {
                  "kind": "StringExpr",
                  "value": "blue",
                },
                "child": Object {
                  "body": Array [
                    Object {
                      "kind": "TextExpr",
                      "value": "233",
                    },
                  ],
                  "closingTag": Object {
                    "kind": "ClosingTagExpr",
                    "tagName": Object {
                      "kind": "IdentityExpr",
                      "name": "span",
                    },
                  },
                  "kind": "JsxExpr",
                  "openingTag": Object {
                    "kind": "OpeningTagExpr",
                    "props": Array [],
                    "tagName": Object {
                      "kind": "IdentityExpr",
                      "name": "span",
                    },
                  },
                },
                "color": Object {
                  "kind": "StringExpr",
                  "value": "red",
                },
                "test": Object {
                  "kind": "ObjectExpr",
                  "props": Object {
                    "color": Object {
                      "kind": "StringExpr",
                      "value": "red",
                    },
                  },
                },
                "width": Object {
                  "kind": "NumberExpr",
                  "value": 100,
                },
              },
            },
          },
          Object {
            "key": Object {
              "kind": "IdentityExpr",
              "name": "id",
            },
            "kind": "PropExpr",
            "value": Object {
              "kind": "StringExpr",
              "value": "233ccc",
            },
          },
          Object {
            "key": Object {
              "kind": "IdentityExpr",
              "name": "class2Name",
            },
            "kind": "PropExpr",
            "value": Object {
              "kind": "StringExpr",
              "value": "qwq123",
            },
          },
        ],
        "tagName": Object {
          "kind": "IdentityExpr",
          "name": "div",
        },
      },
    },
  ],
  "kind": "Program",
}
`;
